### ISSUE TITLE
Upgrade 3.0 Tesla Location tracker not work (China Server)

### ISSUE THREAD

**Issue by BlueBenson** [alandtse/tesla#282]
## Version of the custom_component
3.0
## Configuration

```yaml
Add your logs here.
```

## Describe the bug

Upgrade 3.0 Tesla Location tracker not work (China Server)
Wendy's TeslaX Location tracker was detected at unknown

## Debug log

```text

Add your logs here.

```

**Comment by alandtse** [comment]
Please enable debug logs for tesla and teslajsonpy and check for any errors.

**Comment by alandtse** [comment]
Please also confirm:
1. You've removed and readded the component.
2. You've used the Chinese auth server to login.
3. Other features work.

**Comment by BlueBenson** [comment]
Yes I confirm. I deleted the old component and reinstalled. I used the china auth. Except the GPS other functions all good. Sent from my iPhoneOn Oct 22, 2022, at 18:51, Alan Tse ***@***.***> wrote:﻿
Please also confirm:

You've removed and readded the component.
You've used the Chinese auth server to login.
Other features work.

**Comment by goxofy** [comment]
same problem

China server(tesla.cn)

**Comment by BlueBenson** [comment]
> Please enable debug logs for tesla and teslajsonpy and check for any errors.

2022-10-22 22:20:13.896 DEBUG (MainThread) [custom_components.tesla_custom] Connected to the Tesla API
2022-10-22 22:20:13.897 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:20:18.614 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 4.718 seconds (success: True)
2022-10-22 22:20:18.639 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:20:18.640 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:20:28.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:20:28.000 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:20:38.001 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:20:38.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:20:48.001 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:20:48.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:20:58.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:20:58.000 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:21:08.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:21:08.241 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.241 seconds (success: True)
2022-10-22 22:21:18.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:21:18.000 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.001 seconds (success: True)
2022-10-22 22:21:28.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:21:28.000 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:21:38.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:21:38.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:21:48.001 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:21:48.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:21:58.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:21:58.000 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:22:08.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:22:08.229 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.230 seconds (success: True)
2022-10-22 22:22:18.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:22:18.000 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:22:28.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:22:28.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:22:38.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:22:38.000 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:22:48.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:22:48.000 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:22:58.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:22:58.000 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.001 seconds (success: True)
2022-10-22 22:23:08.001 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:23:08.233 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.233 seconds (success: True)
2022-10-22 22:23:18.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:23:18.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:23:28.002 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:23:28.002 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:23:38.001 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:23:38.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:23:48.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:23:48.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:23:58.001 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:23:58.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:24:08.001 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:24:08.223 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.223 seconds (success: True)
2022-10-22 22:24:18.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:24:18.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:24:28.001 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:24:28.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:24:38.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:24:38.000 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:24:48.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:24:48.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.001 seconds (success: True)
2022-10-22 22:24:58.001 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:24:58.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:25:08.001 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:25:08.225 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.224 seconds (success: True)
2022-10-22 22:25:18.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:25:18.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:25:28.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:25:28.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:25:38.001 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:25:38.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:25:48.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:25:48.000 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.001 seconds (success: True)
2022-10-22 22:25:58.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:25:58.000 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:26:08.001 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:26:08.235 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.234 seconds (success: True)
2022-10-22 22:26:18.001 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:26:18.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:26:28.001 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:26:28.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:26:38.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:26:38.000 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:26:48.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:26:48.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:26:58.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:26:58.000 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:27:08.000 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:27:08.242 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.241 seconds (success: True)
2022-10-22 22:27:18.001 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:27:18.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)
2022-10-22 22:27:28.001 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2022-10-22 22:27:28.001 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 0.000 seconds (success: True)

**Comment by shred86** [comment]
Looking at the logs above, this is what's listed in the "drive_state" key:

````
"drive_state": {
    "corrected_latitude": 24.980499,
    "corrected_longitude": 102.666374,
    "gps_as_of": 1666449682,
    "heading": 5,
    "latitude": 24.983561,
    "longitude": 102.665039,
    "native_latitude": 24.980501,
    "native_location_supported": 1,
    "native_longitude": 102.666374,
    "native_type": "gcj",
    "power": 0,
    "shift_state": null,
    "speed": null,
    "timestamp": 1666449700353
}
````

I'm not too familiar with GCJ-02 but it [appears to be based on WGS-84 and is adding offsets to the latitude and longitude](https://abstractkitchen.com/blog/a-short-guide-to-chinese-coordinate-system/).

Looking at what the Tesla API is returning, there's three different possible values, `corrected_latitude`, `latitude` and `native_latitude`, none of which values match. It also passes the `native_type` of `gcj`.

Are we sure `latitude` or `corrected_latitude` isn't already the converted GCJ-02 coordinates to WGS-84? If so, this would greatly simplify our problem. Additionally, I think we could just check the `native_type` field here instead of checking the auth_domain the user is using.

**Comment by BlueBenson** [comment]
Hi @alandtse @shred86 ,thank you guys so much. I replace the const.py&device_tracker.py from https://github.com/shred86/tesla/tree/china-coords. it works now.

**Comment by shred86** [comment]
Should be fixed in v3.0.1.
